json_serializable/lib/src/enum_utils.dart

```diff
@@ -33,20 +33,50 @@ String? enumValueMapFromType(
 
     dynamic fieldValue;
     if (annotation == null) {
-      fieldValue = encodedFieldName(jsonEnum.fieldRename, fe.name);
+      if (jsonEnum.valueField != null) {
+        // TODO: fieldRename is pointless here!!! At least log a warning!
+
+        final fieldElementType = fe.type.element2 as EnumElement;
+
+        final e = fieldElementType.getField(jsonEnum.valueField!);
+
+        if (e == null || e.isStatic) {
+          throw InvalidGenerationSourceError(
+            '`JsonEnum.valueField` was set to "${jsonEnum.valueField}", but '
+            'that is not a valid, instance field on '
+            '`${typeToCode(targetType)}`.',
+            element: targetType.element2,
+          );
+        }
+
+        final reader = ConstantReader(fe.computeConstantValue());
+        final valueReader = reader.read(jsonEnum.valueField!);
+        if (valueReader.validValueType) {
+          fieldValue = valueReader.literalValue;
+        } else {
+          throw InvalidGenerationSourceError(
+            '`JsonEnum.valueField` was set to "${jsonEnum.valueField}", but '
+            'that field does not have a type of String, int, or null.',
+            element: targetType.element2,
+          );
+        }
+      } else {
+        fieldValue = encodedFieldName(jsonEnum.fieldRename, fe.name);
+      }
```

Comment:
This method was 37 lines, it's now 61 lines. It previously had 2 levels of nesting conditionals, now it has 3 levels.

Consider refactoring to split into smaller chunks.

Reply:
https://github.com/google/json_serializable.dart/pull/1204